services/src/test/scala/cromwell/pipeline/service/ProjectSearchServiceTest.scala

```diff
@@ -0,0 +1,196 @@
+      "return empty result if nothing matched" in {
+        val regexString = ".*UNMATCHED.*"
+        val filter = ByName(RegexpMatch, regexString)
+        val request = ProjectSearchRequest(filter)
+        val result = ProjectSearchResponse(Seq.empty)
+
+        projectSearch.searchProjects(request, userId).map(_ shouldBe result)
+      }
+    }
+
+    "search project by config" should {
+      "return projects with configuration if Exists true" in {
+        val filter = ByConfig(mode = Exists, value = true)
+        val request = ProjectSearchRequest(filter)
+        val result = ProjectSearchResponse(Seq(dummyProject1, dummyProject2))
+
+        projectSearch.searchProjects(request, userId).map(_ shouldBe result)
+      }
+      "return project without configuration if Exists false" in {
+        val filter = ByConfig(mode = Exists, value = false)
+        val request = ProjectSearchRequest(filter)
+        val result = ProjectSearchResponse(Seq(dummyProject3))
+
+        projectSearch.searchProjects(request, userId).map(_ shouldBe result)
+      }
+      "fail with exception if unable to fetch configuration" in {
+        val failedConfigService = ProjectConfigurationServiceTestImpl.withException(new RuntimeException)
+        val projectSearch = createProjectSearchService(projectConfigurationService = failedConfigService)
+
+        val filter = ByConfig(mode = Exists, value = true)
+        val request = ProjectSearchRequest(filter)
+
+        projectSearch.searchProjects(request, userId).failed.map {
+          _ shouldBe InternalError("Failed to fetch configuration due to unexpected internal error")
+        }
+      }
+    }
+
+    "search project by files" should {
+      "return projects with files if Exists true" in {
+        val filter = ByFiles(mode = Exists, value = true)
+        val request = ProjectSearchRequest(filter)
+        val result = ProjectSearchResponse(Seq(dummyProject1, dummyProject2))
+
+        projectSearch.searchProjects(request, userId).map(_ shouldBe result)
+      }
+      "return project without files if Exists false" in {
+        val filter = ByFiles(mode = Exists, value = false)
+        val request = ProjectSearchRequest(filter)
+        val result = ProjectSearchResponse(Seq(dummyProject3))
+
+        projectSearch.searchProjects(request, userId).map(_ shouldBe result)
+      }
+      "fail with exception if unable to fetch configuration" in {
+        val failedFilesService = ProjectFileServiceIdCombinedTestImpl.withException(new RuntimeException)
+        val projectSearch = createProjectSearchService(projectFileService = failedFilesService)
+
+        val filter = ByFiles(mode = Exists, value = true)
+        val request = ProjectSearchRequest(filter)
+
+        projectSearch.searchProjects(request, userId).failed.map {
+          _ shouldBe InternalError("Failed to fetch files due to unexpected internal error")
+        }
+      }
+    }
+
+    "search project with complex filter" should {
```

Comment:
I think it would be useful to add tests on `And` filter that shows that projects that satisfy only some of the conditions won't be included to the result. The only existing test doesn't show that `And` filter can actually exclude anything.

Reply:
Test added

Reply:
I don't see this test. Right now your implementation has a bug. I expect you to write a test that can catch it and then of course fix this bug.

Reply:
That was an embarrassing one. In `projectSearchService` initial Seq of projects was always used inside `doSearch` method.
Test added, `doSearch` fixed